ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp

```diff
@@ -24,339 +25,249 @@
-                                                     in_transform_pad_above,
-                                                     in_transform_dilation_strides);
-
-                    // Simultaneously with iterating I, for the filter we need to iterate the
-                    // coordinate:
-                    //
-                    //   F
-                    //
-                    // over the range (noninclusive on the right):
-                    //
-                    //   (chan_out,0,0,...,0) ->
-                    //     (chan_out+1,
-                    //      chans_in_count,
-                    //      filter_dims_1,
-                    //        ...,
-                    //      filter_dims_n)
-                    //
-                    // with unit stride.
+                }
+            }
 
-                    Shape filter_transform_start(2 + n_spatial_dimensions);
-                    Shape filter_transform_end(2 + n_spatial_dimensions);
+            template <typename INPUT,
+                      typename FILTER,
+                      typename OUTPUT,
+                      typename ACCU = typename widen<OUTPUT>::type>
+            void convolution(const INPUT* in,
+                             const FILTER* f,
+                             OUTPUT* out,
+                             const Shape& in_shape,
+                             const Shape& f_shape,
+                             const Shape& out_shape,
+                             const Strides& strides,
+                             const Strides& dilation,
+                             const CoordinateDiff& pads_begin,
+                             const CoordinateDiff& pads_end)
 
-                    filter_transform_start[filter_out_channel_axis] = out_channel;
-                    filter_transform_end[filter_out_channel_axis] = out_channel + 1;
-                    filter_transform_start[filter_in_channel_axis] = 0;
-                    filter_transform_end[filter_in_channel_axis] = 1;
+            {
+                // this implementation supports 1D, 2D and 3D convolutions
+                NGRAPH_CHECK(in_shape.size() >= 3 && in_shape.size() <= 5,
+                             "Unsupported input rank: ",
+                             in_shape);
 
-                    for (size_t i = 2; i < n_spatial_dimensions + 2; i++)
-                    {
-                        filter_transform_start[i] = 0;
-                        filter_transform_end[i] = filter_shape[i];
-                    }
+                NGRAPH_CHECK(f_shape.size() >= 3 && f_shape.size() <= 5,
+                             "Unsupported kernel rank: ",
+                             f_shape);
 
-                    CoordinateTransform filter_transform(
-                        filter_shape, filter_transform_start, filter_transform_end);
+                auto old_mode = std::fegetround();
+                std::fesetround(FE_TONEAREST);
```

Comment:
@ilyachur one more question regarding old implementation "artifacts". Do you know a reason for this call to std::fesetround() ? Shouldn't floating round mode be the same as in other ops ?

Reply:
@jdanieck I don't know the reason. I think we can try to remove it.

Reply:
Removed in this [commit](https://github.com/openvinotoolkit/openvino/pull/3922/commits/196114166eb3e7812bad73960d5e1d148f4bc80d).